design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+
+#### StorageCluster
+
+The `storageCluster.rook.io/v1alpha2` type provides the general abstraction for a distributed storage cluster.
+The spec for this type will include image version, networking information, placement information, resource usage requests, etc.
+Backend specific fields, such as Ceph configuration, will be included under a field identified by the backend's name.
+Examples of the `StorageCluster` type are included in the [examples](#examples) section.
+
+In addition to the existing support for Ceph, we should consider support for [Minio](https://www.minio.io/) as the next storage backend supported by Rook.
+Minio is a popular distributed object storage solution that could benefit from a custom controller to orchestrate, configure and manage it in cloud-native environments.
+
+#### Additional Types
+
+Rook and the greater community would also benefit from additional types and abstractions.
+We should work on defining those further, but it is out of scope for this design document that is focusing on support for multiple storage backends.
+Some potential ideas for additional types to support in Rook:
+
+* [Snapshots, backup and policy](https://github.com/rook/rook/issues/1552)
+* Quality of Service (QoS), resource consumption (I/O and storage limits)
+
+## Examples
+
+This section contains concrete examples of storage clusters as a user would define them using yaml.
+In addition to the Ceph and Minio examples, we will be considering support for additional storage types in the near future.
+
+`ceph-cluster.yaml`:
+
+```yaml
+apiVersion: rook.io/v1alpha2
+kind: StorageCluster
+metadata:
+ name: ceph
+ namespace: rook-ceph
+spec:
+ version: "rook/ceph:v0.1.0"
+ dataDirHostPath: "/var/lib/rook-ceph"
+ hostNetwork: false
+ placement:
+ - name: mon
+   nodeAffinity:
+   podAffinity:
+   podAntiAffinity:
+   tolerations:
+ resources:
+ - name: osd
+   limits:
+     cpu: "500m"
+     memory: "1024Mi"
+   requests:
+     cpu: "500m"
+     memory: "1024Mi"
+ scope:
+  deviceFilter: "^sd."
+  location:
+  directories:
+  - path: /rook/storage-dir
+  nodes:
+  - name: "nodeA"
+    directories:
+    - path: "/rook/storage-dir"
+  - name: "nodeB"
+    devices:
+    - name: "vdx"
+    - fullpath: "/dev/disk/by-id/abc-123"
+  - name: "machine*"  # wild cards are supported
+    volumeClaimTemplates:
+    - metadata:
+      name: my-pvc-template
+      spec:
+        accessModes: [ "ReadWriteOnce" ]
+        storageClassName: "my-storage-class"
+        resources:
+          requests:
+            storage: "1Gi"
+ ceph:
+   monCount: 3
+   storeConfig:
+     storeType: bluestore
+     databaseSizeMB: 1024
+   metadataDevice: nvme01
+```
+
+`minio-cluster.yaml`:
+
+```yaml
+apiVersion: rook.io/v1alpha2
+kind: StorageCluster
+metadata:
+ name: minio
+ namespace: rook-minio
+spec:
+ version: "rook/minio:v0.1.0"
+ dataDirHostPath: "/var/lib/rook-minio"
+ hostNetwork: false
+ placement:
+ resources:
+ scope:
+   replicas: 4 # use 4 nodes in the cluster
```

Comment:
i'm seeing `replicas` as a common `StorageScope` setting actually.  The general principle of how many storage hosts do you want running, instead of naming the specific nodes to run on, feels like it has applicability to other potential backends.

Reply:
for ceph this would mean that we randomly pick four nodes to start OSDs? What if there are multiple devices on a node? Use all of them for OSDs? I guess that could be a different setting.